web/src/client/http.js

```diff
@@ -38,11 +55,65 @@ class WSClient {
    * @param {URL} url - Websocket URL.
    */
   constructor(url) {
-    this.client = new WebSocket(url.toString());
+    this.url = url.toString();
+
+    this.event_handlers = [];
+    this.close_handlers = [];
+    this.error_handlers = [];
+    this.open_handlers = [];
+    this.reconnectAttempts = 0;
+    this.connect();
+  }
+
+  wsState() {
+    const state = this.client.readyState;
+    if ((state !== SocketStates.CONNECTED) && (this.reconnectAttempts >= MAX_ATTEMPTS)) return SocketStates.UNRECOVERABLE;
+
+    return state;
+  }
+
+  isRecoverable() {
+    return (this.wsState() !== SocketStates.UNRECOVERABLE);
+  }
+
+  isConnected() {
+    return (this.wsState() === SocketStates.CONNECTED);
+  }
+
+  subscribe_to_events() {
+    this.client.onopen = () => {
+      console.log("Connected websocket");
+      this.reconnectAttempts = 0;
+      if (this.timeout !== undefined) clearTimeout(this.timeout);
+
+      return this.dispatchOpenEvent();
+    };
+
     this.client.onmessage = (event) => {
       this.dispatchEvent(event);
     };
-    this.handlers = [];
+
+    this.client.onclose = () => {
+      console.log(`Closed WebSocket`);
+      this.dispatchCloseEvent();
+      this.timeout = setTimeout(() => this.connect(this.reconnectAttempts + 1), ATTEMPT_INTERVAL);
+    };
+
+    this.client.onerror = (e) => {
+      console.error("WebSocket error:", e);
+      this.dispatchErrorEvent();
+    };
+  }
+
+  connect(attempt = 0) {
+    this.reconnectAttempts = attempt;
+    if (attempt > 10) {
```

Comment:
here it should use constant defined above

Reply:
yep, it was temporal and already done ;)